include/swift/SIL/SILInstruction.h

```diff
@@ -4552,6 +4554,32 @@ class FullApplySite : public ApplySite {
   }
 };
 
+// This is defined out of line to work around the fact that this depends on
+// PartialApplyInst being defined, but PartialApplyInst is a subclass of
+// ApplyInstBase, so we can not place ApplyInstBase after it.
+template <class Impl, class Base>
+SILFunction *ApplyInstBase<Impl, Base, false>::getCalleeFunction() const {
+  SILValue Callee = getCallee();
+
+  while (true) {
+    if (auto *FRI = dyn_cast<FunctionRefInst>(Callee)) {
+      return FRI->getReferencedFunction();
+    }
+
+    if (auto *PAI = dyn_cast<PartialApplyInst>(Callee)) {
+      Callee = PAI->getCallee();
+      continue;
+    }
+
+    if (auto *TTTFI = dyn_cast<ThinToThickFunctionInst>(Callee)) {
+      Callee = TTTFI->getCallee();
+      continue;
+    }
+
+    return nullptr;
+  }
+}
+
```

Comment:
> On Feb 22, 2016, at 4:48 PM, eeckstein <notifications@github.com> wrote:
> 
> In include/swift/SIL/SILInstruction.h <https://github.com/apple/swift/pull/1393#discussion_r53720424>:
> 
> > +    }
> > +
> > +    if (auto *PAI = dyn_cast<PartialApplyInst>(Callee)) {
> > +      Callee = PAI->getCallee();
> > +      continue;
> > +    }
> > +
> > +    if (auto *TTTFI = dyn_cast<ThinToThickFunctionInst>(Callee)) {
> > +      Callee = TTTFI->getCallee();
> > +      continue;
> > +    }
> > +
> > +    return nullptr;
> > +  }
> > +}
> > +
> You change the semantics of getCalleeFunction(). I'm not sure if this is safe.
> E.g.:
> %1 = function_ref foo()
> %2 = partial_apply %1()
> %3 = apply %2
> 
> Now it would return foo() if called for the apply instruction %3. I think some optimizations, like the inliner, rely on that getCalleeFunction() just returns a function if it's directly the function_ref operand of the passed apply.
> 
> 

I just discovered that (see the failed build log).

I don't think that there is any reason why this API should have that restriction. We should add a different API that does that.

Michael
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1393/files#r53720424>.
>